File: android/app/src/main/java/com/strainwise/app/ui/browse/DirectoryFilter.kt
package com.strainwise.app.ui.browse

import com.strainwise.app.models.Conditions
import com.strainwise.app.models.StrainProfile
import com.strainwise.app.models.StrainType

/**
 * Search + type / THC / effect filters for the popular
 * catalog. 1:1 port of the iOS `DirectoryFilter.swift` /
 * web `src/components/directory/StrainDirectory.tsx`.
 */
object DirectoryFilter {

    enum class TypeFilter(val label: String, val strainType: StrainType?) {
        All("All types", null),
        Sativa("Sativa", StrainType.Sativa),
        Hybrid("Hybrid", StrainType.Hybrid),
        Indica("Indica", StrainType.Indica);
    }

    enum class ThcBand(val label: String, val rangeLabel: String) {
        Any("Any THC", "no preference"),
        Mild("Mild", "under ~15%"),
        Balanced("Balanced", "~15-22%"),
        Strong("Strong", "above ~22%");

        fun contains(midpoint: Double): Boolean = when (this) {
            Any -> true
            Mild -> midpoint < 15
            Balanced -> midpoint >= 15 && midpoint < 22
            Strong -> midpoint >= 22
        }
    }

    data class EffectBucket(
        val id: String,
        val label: String,
        val keywords: List<String>,
    ) {
        companion object {
            val Relaxing = EffectBucket("relaxed", "Relaxing", listOf("relaxed", "calm", "calming", "soothing"))
            val Sleepy = EffectBucket("sleepy", "Sleepy", listOf("sleepy", "sedated", "drowsy"))
            val Happy = EffectBucket("happy", "Happy", listOf("happy", "euphoric", "uplifted", "giggly"))
            val Focused = EffectBucket("focused", "Focused", listOf("focused", "creative", "aroused"))
            val Energetic = EffectBucket("energetic", "Energetic", listOf("energetic", "tingly", "talkative"))
            val Hungry = EffectBucket("hungry", "Hungry", listOf("hungry", "appetite"))

            val all: List<EffectBucket> = listOf(Relaxing, Sleepy, Happy, Focused, Energetic, Hungry)

            fun named(id: String): EffectBucket? = all.firstOrNull { it.id == id }
        }
    }

    /** Parse a Leafly-style THC range (`"17-24%"`, `"~20%"`, `"<1%"`)
     *  to a numeric midpoint. En-dashes (U+2013) and em-dashes
     *  (U+2014) are treated as hyphens. */
    fun thcMidpoint(range: String?): Double? {
        if (range.isNullOrEmpty()) return null
        val normalized = range
            .replace("\u2013", "-")
            .replace("\u2014", "-")
        val stripped = normalized
            .replace(Regex("[%~\\s<>]"), "")
            .trim()
        if (stripped.isEmpty()) return null

        if (normalized.contains("<")) {
            val digits = stripped.replace(Regex("[^0-9.]"), "")
            val n = digits.toDoubleOrNull() ?: return null
            return maxOf(0.0, n - 0.5)
        }

        val parts = stripped.split("-")
        if (parts.size == 2) {
            val a = parts[0].toDoubleOrNull() ?: return null
            val b = parts[1].toDoubleOrNull() ?: return null
            return (a + b) / 2
        }
        return stripped.toDoubleOrNull()
    }

    fun matches(profile: StrainProfile, bucket: EffectBucket): Boolean {
        val names = (profile.effects ?: emptyList()).map { it.name.lowercase() }.toSet()
        return bucket.keywords.any { names.contains(it) }
    }

    fun matches(
        profile: StrainProfile,
        query: String = "",
        type: TypeFilter = TypeFilter.All,
        thc: ThcBand = ThcBand.Any,
        effectIDs: List<String> = emptyList(),
        ailments: List<String> = emptyList(),
    ): Boolean {
        type.strainType?.let { wanted ->
            if (profile.type != wanted) return false
        }
        val q = query.trim().lowercase()
        if (q.isNotEmpty() && !profile.name.lowercase().contains(q)) return false
        if (thc != ThcBand.Any) {
            val mid = thcMidpoint(profile.thcRange) ?: return false
            if (!thc.contains(mid)) return false
        }
        val buckets = effectIDs.mapNotNull { EffectBucket.named(it) }
        if (buckets.isNotEmpty() && !buckets.all { matches(profile, it) }) return false
        if (ailments.isNotEmpty() && !ailments.all { matchesCondition(it, profile.medicalUses ?: emptyList()) }) {
            return false
        }
        return true
    }

    fun apply(
        profiles: List<StrainProfile>,
        query: String,
        type: TypeFilter,
        thc: ThcBand,
        effectIDs: List<String>,
        ailments: List<String> = emptyList(),
    ): List<StrainProfile> = profiles.filter {
        matches(it, query, type, thc, effectIDs, ailments)
    }

    fun matchesCondition(ailment: String, uses: List<String>): Boolean {
        if (ailment.isEmpty() || uses.isEmpty()) return false
        val keys = Conditions.matchKeys(ailment).map { it.lowercase() }
        return uses.any { use -> keys.contains(use.lowercase()) }
    }
}

File: ios/StrainEase/Directory/DirectoryFilter.swift
import Foundation

/// Search + type / THC / effect filters for the popular catalog.
/// Mirrors `src/components/directory/StrainDirectory.tsx`.
enum DirectoryFilter {
    enum TypeFilter: String, CaseIterable, Identifiable, Sendable {
        case all
        case sativa
        case hybrid
        case indica

        var id: String { rawValue }

        var label: String {
            switch self {
            case .all: "All types"
            case .sativa: "Sativa"
            case .hybrid: "Hybrid"
            case .indica: "Indica"
            }
        }

        var strainType: StrainType? {
            switch self {
            case .all: nil
            case .sativa: .sativa
            case .hybrid: .hybrid
            case .indica: .indica
            }
        }
    }

    enum ThcBand: String, CaseIterable, Identifiable, Sendable {
        case any
        case mild
        case balanced
        case strong

        var id: String { rawValue }

        var label: String {
            switch self {
            case .any: "Any THC"
            case .mild: "Mild"
            case .balanced: "Balanced"
            case .strong: "Strong"
            }
        }

        var rangeLabel: String {
            switch self {
            case .any: "no preference"
            case .mild: "under ~15%"
            case .balanced: "~15–22%"
            case .strong: "above ~22%"
            }
        }

        func contains(_ midpoint: Double) -> Bool {
            switch self {
            case .any: true
            case .mild: midpoint < 15
            case .balanced: midpoint >= 15 && midpoint < 22
            case .strong: midpoint >= 22
            }
        }
    }

    struct EffectBucket: Hashable, Identifiable, Sendable {
        let id: String
        let label: String
        let keywords: [String]

        static let relaxing = EffectBucket(
            id: "relaxed",
            label: "Relaxing",
            keywords: ["relaxed", "calm", "calming", "soothing"]
        )
        static let sleepy = EffectBucket(
            id: "sleepy",
            label: "Sleepy",
            keywords: ["sleepy", "sedated", "drowsy"]
        )
        static let happy = EffectBucket(
            id: "happy",
            label: "Happy",
            keywords: ["happy", "euphoric", "uplifted", "giggly"]
        )
        static let focused = EffectBucket(
            id: "focused",
            label: "Focused",
            keywords: ["focused", "creative", "aroused"]
        )
        static let energetic = EffectBucket(
            id: "energetic",
            label: "Energetic",
            keywords: ["energetic", "tingly", "talkative"]
        )
        static let hungry = EffectBucket(
            id: "hungry",
            label: "Hungry",
            keywords: ["hungry", "appetite"]
        )

        static let all: [EffectBucket] = [
            .relaxing, .sleepy, .happy, .focused, .energetic, .hungry,
        ]

        static func named(_ id: String) -> EffectBucket? {
            all.first { $0.id == id }
        }
    }

    /// Parse Leafly-style range strings (`"17-24%"`, `"~20%"`, `"<1%"`)
    /// to a numeric midpoint. En-dashes from the local catalog are
    /// treated as hyphens so `"17–24%"` matches the web hyphen form.
    static func thcMidpoint(_ range: String?) -> Double? {
        guard let range, !range.isEmpty else { return nil }
        let normalized = range
            .replacingOccurrences(of: "–", with: "-")
            .replacingOccurrences(of: "—", with: "-")
        let stripped = normalized.replacingOccurrences(
            of: #"[%~\s<>]"#,
            with: "",
            options: .regularExpression
        ).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !stripped.isEmpty else { return nil }

        if normalized.contains("<") {
            let digits = stripped.replacingOccurrences(
                of: #"[^0-9.]"#,
                with: "",
                options: .regularExpression
            )
            guard let n = Double(digits) else { return nil }
            return max(0, n - 0.5)
        }

        let parts = stripped.split(separator: "-", omittingEmptySubsequences: false)
        if parts.count == 2, let a = Double(parts[0]), let b = Double(parts[1]) {
            return (a + b) / 2
        }
        return Double(stripped)
    }

    static func matches(_ profile: StrainProfile, bucket: EffectBucket) -> Bool {
        let names = Set((profile.effects ?? []).map { $0.name.lowercased() })
        return bucket.keywords.contains { names.contains($0) }
    }

    static func matches(
        _ profile: StrainProfile,
        query: String = "",
        type: TypeFilter = .all,
        thc: ThcBand = .any,
        effectIDs: [String] = [],
        ailments: [String] = []
    ) -> Bool {
        if let wanted = type.strainType, profile.type != wanted {
            return false
        }
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if !q.isEmpty, !profile.name.lowercased().contains(q) {
            return false
        }
        if thc != .any {
            guard let mid = thcMidpoint(profile.thcRange), thc.contains(mid) else {
                return false
            }
        }
        let buckets = effectIDs.compactMap(EffectBucket.named)
        if !buckets.isEmpty, !buckets.allSatisfy({ matches(profile, bucket: $0) }) {
            return false
        }
        if !ailments.isEmpty,
           !ailments.allSatisfy({ matchesCondition(ailment: $0, uses: profile.medicalUses ?? []) }) {
            return false
        }
        return true
    }

    static func apply(
        to profiles: [StrainProfile],
        query: String,
        type: TypeFilter,
        thc: ThcBand,
        effectIDs: [String],
        ailments: [String] = []
    ) -> [StrainProfile] {
        profiles.filter {
            matches(
                $0,
                query: query,
                type: type,
                thc: thc,
                effectIDs: effectIDs,
                ailments: ailments
            )
        }
    }

    /// True when at least one of the strain's reported uses matches the
    /// curated ailment (case-insensitive, alias-aware). Mirrors
    /// `src/lib/strain-ui.ts#matchesCondition` so web + iOS read the
    /// same set of conditions.
    static func matchesCondition(ailment: String, uses: [String]) -> Bool {
        guard !ailment.isEmpty, !uses.isEmpty else { return false }
        let keys = Conditions.matchKeys(for: ailment).map { $0.lowercased() }
        return uses.contains { use in
            keys.contains(use.lowercased())
        }
    }
}
